### DISCUSSION TITLE
Insufficient privilege: 7 ERROR:  permission denied for table oc_appconfig  after migration

### DISCUSSION THREAD

**Discussion by Leptopoda** [nextcloud/all-in-one#2564] [Questions] [answered]
Hi there,
I've followed the migration instructions to migrate from a custom mysql docker compose to AIO.
I've tried a few times and it always fails with the given error message:
```
Next Doctrine\DBAL\Exception\DriverException: An exception occurred while executing a query: SQLSTATE[42501]: Insufficient privilege: 7 ERROR:  permission denied for table oc_appconfig in /var/www/html/3rdparty/doctrine/dbal/src/Driver/API/PostgreSQL/ExceptionConverter.php:91

ace:
#0 /var/www/html/3rdparty/doctrine/dbal/src/Connection.php(1814): Doctrine\DBAL\Driver\API\PostgreSQL\ExceptionConverter->convert(Object(Doctrine\DBAL\Driver\PDO\Exception), Object(Doctrine\DBAL\Query))
#1 /var/www/html/3rdparty/doctrine/dbal/src/Connection.php(1749): Doctrine\DBAL\Connection->handleDriverException(Object(Doctrine\DBAL\Driver\PDO\Exception), Object(Doctrine\DBAL\Query))
#2 /var/www/html/3rdparty/doctrine/dbal/src/Connection.php(1055): Doctrine\DBAL\Connection->convertExceptionDuringQuery(Object(Doctrine\DBAL\Driver\PDO\Exception), 'SELECT * FROM "...', Array, Array)
#3 /var/www/html/lib/private/DB/Connection.php(262): Doctrine\DBAL\Connection->executeQuery('SELECT * FROM "...', Array, Array, NULL)
#4 /var/www/html/3rdparty/doctrine/dbal/src/Query/QueryBuilder.php(345): OC\DB\Connection->executeQuery('SELECT * FROM "...', Array, Array)
#5 /var/www/html/lib/private/DB/QueryBuilder/QueryBuilder.php(280): Doctrine\DBAL\Query\QueryBuilder->execute()
#6 /var/www/html/lib/private/AppConfig.php(418): OC\DB\QueryBuilder\QueryBuilder->execute()
#7 /var/www/html/lib/private/AppConfig.php(184): OC\AppConfig->loadConfigValues()
#8 /var/www/html/lib/private/AppConfig.php(374): OC\AppConfig->getApps()
#9 /var/www/html/lib/private/legacy/OC_App.php(976): OC\AppConfig->getValues(false, 'installed_versi...')
#10 /var/www/html/lib/private/Server.php(729): OC_App::getAppVersions()
#11 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\Server->OC\{closure}(Object(OC\Server))
#12 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#13 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OC\\Memcache\\Fac...')
#14 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OC\\Memcache\\Fac...', true)
#15 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OC\\Memcache\\Fac...')
#16 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(193): OC\AppFramework\Utility\SimpleContainer->get('OC\\Memcache\\Fac...')
#17 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(OC\Server))
#18 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(118): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#19 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OCP\\ICacheFacto...')
#20 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OCP\\ICacheFacto...', true)
#21 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OCP\\ICacheFacto...')
#22 /var/www/html/lib/private/Server.php(1110): OC\AppFramework\Utility\SimpleContainer->get('OCP\\ICacheFacto...')
#23 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\Server->OC\{closure}(Object(OC\Server))
#24 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#25 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OCP\\Lock\\ILocki...')
#26 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OCP\\Lock\\ILocki...', true)
#27 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OCP\\Lock\\ILocki...')
#28 /var/www/html/lib/private/Server.php(2065): OC\AppFramework\Utility\SimpleContainer->get('OCP\\Lock\\ILocki...')
#29 /var/www/html/lib/private/Files/View.php(119): OC\Server->getLockingProvider()
#30 /var/www/html/lib/private/Server.php(464): OC\Files\View->__construct()
#31 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(171): OC\Server->OC\{closure}(Object(OC\Server))
#32 /var/www/html/3rdparty/pimple/pimple/src/Pimple/Container.php(122): OC\AppFramework\Utility\SimpleContainer->OC\AppFramework\Utility\{closure}(Object(Pimple\Container))
#33 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(138): Pimple\Container->offsetGet('OC\\Files\\Node\\H...')
#34 /var/www/html/lib/private/ServerContainer.php(171): OC\AppFramework\Utility\SimpleContainer->query('OC\\Files\\Node\\H...', true)
#35 /var/www/html/lib/private/AppFramework/Utility/SimpleContainer.php(65): OC\ServerContainer->query('OC\\Files\\Node\\H...')
#36 /var/www/html/lib/private/Server.php(1467): OC\AppFramework\Utility\SimpleContainer->get('OC\\Files\\Node\\H...')
#37 /var/www/html/lib/base.php(625): OC\Server->boot()
#38 /var/www/html/lib/base.php(1167): OC::init()
#39 /var/www/html/console.php(48): require_once('/var/www/html/l...')
#40 /var/www/html/occ(11): require_once('/var/www/html/c...')
```

Although looking at the dumped pg DB it looks like this shouldn't be an issue:
```SQL
--
-- Name: oc_appconfig; Type: TABLE; Schema: public; Owner: nextcloud-aio
--

CREATE TABLE public.oc_appconfig (
    appid character varying(32) DEFAULT ''::character varying NOT NULL,
    configkey character varying(64) DEFAULT ''::character varying NOT NULL,
    configvalue text
);


ALTER TABLE public.oc_appconfig OWNER TO "nextcloud-aio";
```

In case of interest I just added a pg docker container to my existing compose like:
```yaml
  db_pg:
    container_name: nextcloud_db_pg
    image: postgres
    networks:
      - default
    restart: unless-stopped
    environment:
      - POSTGRES_HOST=nextcloud_db_pg
      - POSTGRES_DB=nextcloud
      - POSTGRES_USER=nextcloud-aio
      - POSTGRES_PASSWORD=xxxxxxxxxxx
    volumes:
      - pgsql:/var/lib/pgsql/data
```
migrated to pg using occ and then went on with the aio migration. The instance works after the migration thus I suspect the error lies somewhere on the aio side.

Thanks for the help :)

**Comment by szaimen** [comment]
Hi, did you make sure that the imported Nextcloud did have the exact version before and after?

**Comment by Leptopoda** [comment]
That's probably the case aio: `26.0.2.0` old instance: `26.0.1.1`.

Thanks for the help. I'll try the migration again later.

**Comment by Leptopoda** [comment]
Ok I just tried again from the beginning (fresh docker host ...)
I updated the existing instance to the latest 26.0.2rc1 an still get the same error.

do you have any idea on how I can fix this? Maybe force giving it the privileges (my sql experience is best described as basic).
Or is this some other bug I'm hitting.

**Comment by szaimen** [comment]
Did you follow https://github.com/nextcloud/all-in-one/blob/main/migration.md#migrate-the-files-and-the-database step-for-step?

**Comment by Leptopoda** [comment]
More or less.

I did the initial mysql->postgres migration (step 3) differently. I just spun up a postgres container as seen above and run the occ command.
As the migration completed normally and the old instance is running fine after migration I'd argue that this isn't the problem.

Other than that I followed it step by step.

**Comment by szaimen** [comment, marked as answer]
>  (step 3)  differently. I just spun up a postures container as seen above and run the occ command.

This likely explains the problem. You can use a container but please use the in the migration docs mentioned commands to create the postgres user and database.

**Comment by jakobroehrl** [comment]
Hi, I have the same prop. 
Version from and to are the same. I followed the migration guide and try to migrate only the db for testing. On the from machine I run postgres already.

Error: Insufficient privilege: 7 ERROR:  permission denied for table oc_appconfig

Dump File:

```
CREATE TABLE public.oc_appconfig (
    appid character varying(32) DEFAULT ''::character varying NOT NULL,
    configkey character varying(64) DEFAULT ''::character varying NOT NULL,
    configvalue text
);
ALTER TABLE public.oc_appconfig OWNER TO nextcloud;
```

Database Config
![grafik](https://github.com/nextcloud/all-in-one/assets/24530680/dab3da50-e6c6-4e4e-89a3-44882b5c266f)

**Reply by Leptopoda** [reply]
This was the odd thing with my issue.
I did the mysql => postgres migration "my way" and the instance was doing fine for days.

Only the postgres -> postgres aio thing didn't work.
Maybe there is a postgres command that can change the problematic permissions.

And this should at least be added to the docs.

**Reply by szaimen** [reply]
I have the feeling you did not create the database correctly while doing the migration. Please use the exact commands from the migration docs to create it. This should usually make it work.

**Comment by jakobroehrl** [comment]
Thanks for your reply. I tested it again:

1. export dump old server: sudo -Hiu postgres pg_dump nextcloud  > ./database-dump.sql
2. check nc versions
3. stop containers
4. change local::/mnt/ncdata/ and owner in the dumpfile
5. copy dump to new server
6. sudo docker run --rm --volume nextcloud_aio_database_dump:/mnt/data:rw alpine rm /mnt/data/database-dump.sql
7. sudo docker cp database-dump.sql nextcloud-aio-database:/mnt/data/
8. sudo docker run --rm --volume nextcloud_aio_database_dump:/mnt/data:rw alpine chmod 777 /mnt/data/database-dump.sql
9. sudo docker run --rm --volume nextcloud_aio_database_dump:/mnt/data:rw alpine rm /mnt/data/initial-cleanup-done
10. edit config file
11. start containers

Insufficient privilege: 7 ERROR: permission denied for table oc_appconfig

**Reply by szaimen** [reply]
Can you post the database container logs here from when it starts up and imports the dump?

**Reply by jakobroehrl** [reply]
sure

Restoring from database dump.
+ rm -rf '/var/lib/postgresql/data/*'
+ export PGPORT=11000
+ PGPORT=11000
+ psql -d postgresql://oc_nextcloud:aca66894dabf60ff85092d42d9b57504315f97a338b9fdcc@localhost:11000/nextcloud_database -c 'select now()'
+ exec docker-entrypoint.sh postgres
psql: error: connection to server at "localhost" (127.0.0.1), port 11000 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (::1), port 11000 failed: Address not available
	Is the server running on that host and accepting TCP/IP connections?
+ echo 'Waiting for the database to start.'
Waiting for the database to start.

5
chmod: /var/run/postgresql: Operation not permitted
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /var/lib/postgresql/data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... Europe/Berlin
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... sh: locale: not found
2023-08-31 13:34:50.350 CEST [34] WARNING:  no usable system locales were found
ok
syncing data to disk ... ok

initdb: warning: enabling "trust" authentication for local connections
initdb: hint: You can change this by editing pg_hba.conf or using the option -A, or --auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data -l logfile start

waiting for server to start....2023-08-31 13:34:50.818 CEST [40] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, 64-bit
2023-08-31 13:34:50.818 CEST [40] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.11000"
2023-08-31 13:34:50.826 CEST [43] LOG:  database system was shut down at 2023-08-31 13:34:50 CEST
2023-08-31 13:34:50.837 CEST [40] LOG:  database system is ready to accept connections
 done
server started
CREATE DATABASE


/usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/init-user-db.sh
+ touch /mnt/data/initialization.failed
+ psql -v ON_ERROR_STOP=1 --username nextcloud --dbname nextcloud_database
CREATE ROLE
ALTER DATABASE
GRANT
GRANT
+ rm /mnt/data/initialization.failed

x

waiting for server to shut down....2023-08-31 13:34:50.993 CEST [40] LOG:  received fast shutdown request
2023-08-31 13:34:50.994 CEST [40] LOG:  aborting any active transactions
2023-08-31 13:34:50.996 CEST [40] LOG:  background worker "logical replication launcher" (PID 46) exited with exit code 1
2023-08-31 13:34:50.996 CEST [41] LOG:  shutting down
2023-08-31 13:34:50.997 CEST [41] LOG:  checkpoint starting: shutdown immediate
2023-08-31 13:34:51.021 CEST [41] LOG:  checkpoint complete: wrote 927 buffers (5.7%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.016 s, sync=0.006 s, total=0.025 s; sync files=257, longest=0.001 s, average=0.001 s; distance=4227 kB, estimate=4227 kB
2023-08-31 13:34:51.027 CEST [40] LOG:  database system is shut down
 done
server stopped

PostgreSQL init process complete; ready for start up.

2023-08-31 13:34:51.111 CEST [16] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, 64-bit
2023-08-31 13:34:51.111 CEST [16] LOG:  listening on IPv4 address "0.0.0.0", port 11000
2023-08-31 13:34:51.111 CEST [16] LOG:  listening on IPv6 address "::", port 11000
2023-08-31 13:34:51.112 CEST [16] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.11000"
2023-08-31 13:34:51.114 CEST [61] LOG:  database system was shut down at 2023-08-31 13:34:51 CEST
2023-08-31 13:34:51.116 CEST [16] LOG:  database system is ready to accept connections
+ psql -d postgresql://oc_nextcloud:aca66894dabf60ff85092d42d9b57504315f97a338b9fdcc@localhost:11000/nextcloud_database -c 'select now()'
              now              
-------------------------------
 2023-08-31 13:34:54.908464+02
(1 row)

+ GREP_STRING='Name: oc_appconfig; Type: TABLE; Schema: public; Owner:'
+ grep -q 'Name: oc_appconfig; Type: TABLE; Schema: public; Owner:' /mnt/data/database-dump.sql
++ ++ grep 'Name: oc_appconfig; Type: TABLE; Schema: public; Owner:' /mnt/data/database-dump.sql
grep ++ -oP 'Owner:.*$'
sed 's|Owner:||;s| ||g'
+ DB_OWNER=ncadmin
+ '[' ncadmin = nextcloud ']'
+ '[' ncadmin '!=' oc_nextcloud ']'
+ DIFFERENT_DB_OWNER=1
+ psql -v ON_ERROR_STOP=1 --username nextcloud --dbname nextcloud_database
CREATE ROLE
ALTER DATABASE
GRANT
GRANT
+ echo 'Restoring the database from database dump'
Restoring the database from database dump
+ psql nextcloud_database -U nextcloud
SET
SET
SET
SET
SET

fig 

----
 
(1 row)

SET
SET
SET
SET
SET
SET

ABLE
ALTER TABLE
...

**Reply by szaimen** [reply]
All right and what comes at the end? Is there a failure or something?

**Reply by jakobroehrl** [reply]
...
+ '[' -n 1 ']'
+ psql -v ON_ERROR_STOP=1 --username nextcloud --dbname nextcloud_database
ALTER DATABASE
REASSIGN OWNED
+ pg_ctl stop -m fast
2023-08-31 14:07:57.429 CEST [15] LOG:  received fast shutdown request
waiting for server to shut down....2023-08-31 14:07:57.430 CEST [15] LOG:  aborting any active transactions
2023-08-31 14:07:57.434 CEST [15] LOG:  background worker "logical replication launcher" (PID 63) exited with exit code 1

31 14:07:57.650 CEST [58] LOG:  checkpoint complete: wrote 1165 buffers (7.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=4.194 s, sync=0.196 s, total=4.404 s; sync files=484, longest=0.046 s, average=0.001 s; distance=535414 kB, estimate=535414 kB
2023-08-31 14:07:57.650 CEST [58] LOG:  shutting down
2023-08-31 14:07:57.651 CEST [58] LOG:  checkpoint starting: shutdown immediate
2023-08-31 14:07:57.689 CEST [58] LOG:  checkpoint complete: wrote 5795 buffers (35.4%); 0 WAL file(s) added, 0 removed, 2 recycled; write=0.034 s, sync=0.002 s, total=0.039 s; sync files=31, longest=0.001 s, average=0.001 s; distance=31626 kB, estimate=485035 kB
2023-08-31 14:07:57.714 CEST [15] LOG:  database system is shut down
 done
server stopped
+ export PGPORT=5432
+ PGPORT=5432

x
Setting max connections...
chmod: /var/run/postgresql: Operation not permitted

PostgreSQL Database directory appears to contain a database; Skipping initialization

2023-08-31 14:07:57.773 CEST [111] LOG:  starting PostgreSQL 15.3 on x86_64-pc-linux-musl, compiled by gcc (Alpine 12.2.1_git20220924-r10) 12.2.1 20220924, 64-bit
2023-08-31 14:07:57.773 CEST [111] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-08-31 14:07:57.773 CEST [111] LOG:  listening on IPv6 address "::", port 5432
2023-08-31 14:07:57.774 CEST [111] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-08-31 14:07:57.779 CEST [121] LOG:  database system was shut down at 2023-08-31 14:07:57 CEST
2023-08-31 14:07:57.783 CEST [111] LOG:  database system is ready to accept connections
2023-08-31 14:08:04.173 CEST [128] ERROR:  permission denied for table oc_appconfig
2023-08-31 14:08:04.173 CEST [128] STATEMENT:  SELECT * FROM "oc_appconfig"

**Reply by szaimen** [reply]
Okay, and did you use this initially when creating the postgres db?
![image](https://github.com/nextcloud/all-in-one/assets/42591237/03911420-3b11-4f51-bb4a-8b58b4eb5d67)

**Reply by jakobroehrl** [reply]
no, because I already use postgres I skiped point 3 of https://github.com/nextcloud/all-in-one/blob/main/migration.md#migrate-the-files-and-the-database

I directly started with point 4, the export

**Reply by szaimen** [reply]
All right, then before doing the db dump possibly these two querys would have been needed:
```
GRANT ALL PRIVILEGES ON DATABASE $PG_DATABASE TO $PG_USER;
GRANT ALL PRIVILEGES ON SCHEMA public TO $PG_USER;
```

**Reply by szaimen** [reply]
I guess we will need to update the migration docs with that.

**Reply by szaimen** [reply]
Can you try that out and report back if it helps?

**Reply by jakobroehrl** [reply]
Unfortunately it does not solve the problem, same error messages.

**Reply by jakobroehrl** [reply]
Btw: How can I use psql within AIO?

**Reply by szaimen** [reply]
Another idea: did you make sure that the nextcloud version from aio and the migrated version match exactly? So currently 27.0.2?

**Reply by jakobroehrl** [reply]
Hi, 
from (installed like here: https://www.c-rieger.de/nextcloud-installationsanleitung/)
![image](https://github.com/nextcloud/all-in-one/assets/24530680/6a1d6c00-f2d9-41bf-be65-77f0257664ff)
to (fresh AIO install)
![image](https://github.com/nextcloud/all-in-one/assets/24530680/7de71d7c-e66b-4147-8234-ec35d46f9fd0)

**Reply by jakobroehrl** [reply]
How can I access the database via psql?

**Reply by rhaamo** [reply]
fwiw I had to change every `OWNER TO nextcloud` to `OWNER TO ncadmin` for nextcloud to work in the dump before copying it into the container.

**Reply by FaySmash** [reply]
I opened a [pull request](https://github.com/nextcloud/all-in-one/pull/4279) with that important missing step.

**Reply by FaySmash** [reply]
Also, if this is still of interest, you can execute psql with:

```
docker exec -it nextcloud-aio-database /bin/bash
psql -U nextcloud nextcloud_database
```

**Comment by saucy-horse** [comment]
very new to servers, nextcloud, computer science in general so apologies if this is only a solution that helped me.

I was running into the same issue: 
_"7 ERROR: permission denied for table oc_migration"_

I'm running dietpi on a Raspberry pi 3 B

I'm setting up nextcloud for the first time so not sure if this is even the same issue in question... but,

I was able to find the file "MySQL.php" in: /home/pi/nextcloud/lib/private/Setup was not granting permission to the dbuser - oc_oc_admin.

I ran: chmod 777 /home/pi/nextcloud/lib/private/Setup/MySQL.php

Granting any user permission to this file which sorted the '7 ERROR..." and I was able to get through to nextcloud after that.

Again, I might be way off base with the original issue that was happening here but if I had seen the above code in my internet trawling I would have saved myself an afternoon. Hope this helps someone in a similar boat to me.
